examples_tests/61.BlitFilterTest/main.cpp

```diff
@@ -0,0 +1,1561 @@
+	auto imageBuffer = core::make_smart_refctd_ptr<asset::ICPUBuffer>(bufferSize);
+	core::smart_refctd_ptr<ICPUImage> image = ICPUImage::create(std::move(imageParams));
+	image->setBufferAndRegions(core::smart_refctd_ptr(imageBuffer), imageRegions);
+
+	if (fillWithTestData)
+	{
+		double pixelValueUpperBound = 20.0;
+		if (asset::isNormalizedFormat(format))
+			pixelValueUpperBound = 1.00000000001;
+
+		std::uniform_real_distribution<double> dist(0.0, pixelValueUpperBound);
+		std::mt19937 prng;
+
+		uint8_t* bytePtr = reinterpret_cast<uint8_t*>(image->getBuffer()->getPointer());
+		for (uint64_t k = 0u; k < dims[2]; ++k)
+		{
+			for (uint64_t j = 0u; j < dims[1]; ++j)
+			{
+				for (uint64_t i = 0; i < dims[0]; ++i)
+				{
+					double decodedPixel[4] = { 0 };
+					for (uint32_t ch = 0u; ch < asset::getFormatChannelCount(format); ++ch)
+						decodedPixel[ch] = dist(prng);
+
+					const uint64_t pixelIndex = (k * dims[1] * dims[0]) + (j * dims[0]) + i;
+					asset::encodePixelsRuntime(format, bytePtr + pixelIndex * asset::getTexelOrBlockBytesize(format), decodedPixel);
+				}
+			}
+		}
+	}
+
+	return image;
+}
+
+static inline asset::IImageView<asset::ICPUImage>::E_TYPE getImageViewTypeFromImageType_CPU(const asset::IImage::E_TYPE type)
+{
+	switch (type)
+	{
+	case asset::IImage::ET_1D:
+		return asset::ICPUImageView::ET_1D;
+	case asset::IImage::ET_2D:
+		return asset::ICPUImageView::ET_2D;
+	case asset::IImage::ET_3D:
+		return asset::ICPUImageView::ET_3D;
+	default:
+		__debugbreak();
+		return static_cast<asset::IImageView<asset::ICPUImage>::E_TYPE>(0u);
+	}
+}
+
+static inline video::IGPUImageView::E_TYPE getImageViewTypeFromImageType_GPU(const video::IGPUImage::E_TYPE type)
+{
+	switch (type)
+	{
+	case video::IGPUImage::ET_1D:
+		return video::IGPUImageView::ET_1D;
+	case video::IGPUImage::ET_2D:
+		return video::IGPUImageView::ET_2D;
+	case video::IGPUImage::ET_3D:
+		return video::IGPUImageView::ET_3D;
+	default:
+		__debugbreak();
+		return static_cast<video::IGPUImageView::E_TYPE>(0u);
+	}
+}
+
+class CBlitFilter
+{
+private:
+	struct alignas(16) vec3_aligned
+	{
+		float x, y, z;
+	};
+
+	struct alignas(16) uvec3_aligned
+	{
+		uint32_t x, y, z;
+	};
+
+public:
+	static constexpr uint32_t NBL_GLSL_DEFAULT_WORKGROUP_DIM = 16u;
+	static constexpr uint32_t NBL_GLSL_DEFAULT_WORKGROUP_SIZE = 256u;
+	static constexpr uint32_t NBL_GLSL_DEFAULT_BIN_COUNT = 256u;
+
+	struct alpha_test_push_constants_t
+	{
+		float referenceAlpha;
+	};
+
+	struct blit_push_constants_t
+	{
+		uvec3_aligned inDim;
+		uvec3_aligned outDim;
+		vec3_aligned negativeSupport;
+		vec3_aligned positiveSupport;
+		uvec3_aligned windowDim;
+		uvec3_aligned phaseCount;
+		uint32_t windowsPerWG;
+		uint32_t axisCount;
+	};
+
+	struct normalization_push_constants_t
+	{
+		uvec3_aligned outDim;
+		uint32_t inPixelCount;
+		float referenceAlpha;
+	};
```

Comment:
I propose a a vec3 definition (like the other utilities do) + `#include "nbl/builtin/glsl/blit/parameters.glsl"` whose content would be:

```glsl
#ifndef _NBL_GLSL_BLIT_NORMALIZATION_PARAMETERS_INCLUDED_
#define _NBL_GLSL_BLIT_NORMALIZATION_PARAMETERS_INCLUDED_

#ifdef __cplusplus
#define uint32_t
#endif
struct nbl_glsl_blit_parameters_t
{
    uvec3 outDim;
    float referenceAlpha;
    vec3 fScale;
    uint inPixelCount;
    vec3 negativeSupport;
    uint padding0;
    uvec3 windowDim;
    uint padding0;
    uvec3 phaseCount;
    uint windowsPerWG;
};
#ifdef __cplusplus
#undef uint
#endif

#endif
```

you don't seem to use `positiveSupport` nor `axisCount`

Reply:
you're not going to do much harm by wasting some PC space in the lighter dispatches, you might even do some good w.r.t. pipeline layout compatibility and descriptor set matching (since PC ranges must match first).

Reply:
Resolved.